crypto/fipsmodule/rand/rand.c

```diff
@@ -470,42 +488,89 @@ void RAND_bytes_with_additional_data(uint8_t *out, size_t out_len,
+        get_entropy_with_no_fast_rdrand(additional_data, sizeof(additional_data));
+      }
+
+      // From now on enforce defending against ube's. This is needed if we need
+      // to reenter the loop for snapsafe.
+      ube_detected_draw_entropy = 1;
+
+      // Update cached generation numbers. This avoids an unnecessary defense
+      // next time we enter |RAND_bytes_with_additional_data|.
+      state->fork_generation = fork_generation;
+      state->snapsafe_generation = snapsafe_generation;
     }
 
-    if (!CTR_DRBG_generate(&state->drbg, out, todo, additional_data,
-                           first_call ? sizeof(additional_data) : 0)) {
-      abort();
+#if defined(BORINGSSL_FIPS)
+    CRYPTO_STATIC_MUTEX_lock_read(state_clear_all_lock_bss_get());
+#endif
+
+    int first_call = 1;
+    while (out_len > 0) {
+      size_t todo = out_len;
+      if (todo > CTR_DRBG_MAX_GENERATE_LENGTH) {
+        todo = CTR_DRBG_MAX_GENERATE_LENGTH;
+      }
+
+      if (!CTR_DRBG_generate(&state->drbg, out, todo, additional_data,
+                             first_call ? sizeof(additional_data) : 0)) {
+        abort();
+      }
+
+      out += todo;
+      out_len -= todo;
+      // Though we only check before entering the loop, this cannot add enough to
+      // overflow a |size_t|.
+      state->calls++;
+      first_call = 0;
     }
 
-    out += todo;
-    out_len -= todo;
-    // Though we only check before entering the loop, this cannot add enough to
-    // overflow a |size_t|.
-    state->calls++;
-    first_call = 0;
-  }
+#if defined(BORINGSSL_FIPS)
+    CRYPTO_STATIC_MUTEX_unlock_read(state_clear_all_lock_bss_get());
+#endif
 
+    // If a process forks, only the thread initiating the fork will be
+    // duplicated. Hence, other forking threads will not affect the currently
+    // executing thread and we can simply check for fork-type ube's once, before
+    // generating randomness. However, this is not true for snapsafe-type ube's,
+    // which can be caused by e.g. process-external events. For example, a VM
+    // snapshot can be taken after the first snapsafe-type ube check. To guard
+    // against such events, we perform a second snapsafe-type ube check.
+    // Why have the first check? The time window in which the snapsafe-type ube
+    // could occur between the first check and returning the randomness is very
+    // small. But, the time window between now and potentially reentering
+    // |RAND_bytes_with_additional_data|, where snapsafe could also be violated,
+    // is much bigger in size.
+    snapsafe_status = CRYPTO_get_snapsafe_generation(&snapsafe_generation);
+  } while (state->snapsafe_generation != snapsafe_generation);
+
+  // Unlikely, but if we use an ephemeral drbg state, it must be released. There
+  // is a tiny time window here where snapsafe-type ube's could go undetected.
   if (state == &stack_state) {
+    // There is no need to mutex. |rand_thread_state_clear_all| simply cleanses
```

Comment:
Where is `rand_thread_state_clear_all` called? I see only a call to `CTR_DRBG_clear` which calls `OPENSSL_cleanse`.

Reply:
`rand_thread_state_clear_all` is defined with the functional attribute `destructor`. This is supposed to be called automatically at the end of main or other process termination calls.

I guess the original authors concern was that this races with some thread still drawing and outputting randomness to an application.

I don't know enough about how the `destructor` functional attribute is actually implemented to diverge from that viewpoint. But in this particular code-line, there is no race concern.